# Preparation Data

Tahap ini dimulai dengan import library yang akan digunakan untuk menganalisis data

In [1]:
# Library yang akan digunakan
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

kemudian buat variabel 'Data' untuk membaca file yang masih berantakan

In [2]:
Data = pd.read_csv('dirty_cafe_sales.csv')
Data

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11
...,...,...,...,...,...,...,...,...
9995,TXN_7672686,Coffee,2,2.0,4.0,NaN,UNKNOWN,2023-08-30
9996,TXN_9659401,NaN,3,NaN,3.0,Digital Wallet,NaN,2023-06-02
9997,TXN_5255387,Coffee,4,2.0,8.0,Digital Wallet,NaN,2023-03-02
9998,TXN_7695629,Cookie,3,NaN,3.0,Digital Wallet,NaN,2023-12-02


# Duplicate

cek apakah terdapat data yang duplicate. Jika ada maka hitung jumlah data yang duplicate

In [3]:
Data1 = Data.copy()

duplikat_data = Data1.duplicated().sum()
print(f'{duplikat_data}')

0


Fungsi 'duplikat_data' menampilkan hasil data duplicate adalah 0. Jadi tidak ada data duplicate

# Data Cleaning & Missing Values

.info() digunakan untuk menampilkan feature(kolom), jumlah data non-null(tidak kosong/bernilai), serta type data yang ada pada file

In [4]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


selanjutnya dilakukan **data cleaning & missing values pada masing-masing feature nya** karena dari info() sebelumnya **terdapat beberapa data null(kosong) dan juga  beberapa type data yang tidak sesuai dengan feature nya** contoh seperti 'Transaction Date' yang seharusnya menggunakan tanggal(datetime) bukan object

## Item

Langkah pertama pada feature item dilakukan pengecekan data seperti data unique dan total missing values

In [5]:
unikItem = Data1['Item'].unique()
misvalItem = Data1['Item'].isna().sum()

print('Item Feauture')
print(f'Data Unik: \n{unikItem}\n')
print(f'Total Missing Value: \n{misvalItem}')

Item Feauture
Data Unik: 
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' nan
 'ERROR' 'Juice' 'Tea']

Total Missing Value: 
333


kemudian membuat sebuah variabel untuk 
1. mengisi data null dengan 'UNKNOWN' pada kolom 'Item'
2. menggabungkan baris yang berisi 'UNKNOWN' dan 'ERROR' pada kolom 'Item' menjadi 'Unknown' agar mudah untuk di analisis selanjutnya

In [6]:
misvalItem1 = Data1['Item'] = Data1['Item'].fillna('UNKNOWN')
Data1['Item'] = Data1['Item'].apply(
    lambda x:'Unknown' if str(x) in ['UNKNOWN','ERROR']
    else str(x)
)

print('Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown\n')
print('Data setelah di diisi dan digabungkan')
Data1['Item'].unique()

Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown

Data setelah di diisi dan digabungkan


array(['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'Unknown',
       'Sandwich', 'Juice', 'Tea'], dtype=object)

cek apakah masih ada data null

In [7]:
Data1['Item'].isna().sum()

np.int64(0)

variabel 'total_Item' digunakan untuk menghitung jumlah data pada masing-masing Itemnya

In [8]:
total_Item = Data1['Item'].value_counts()
print(f'Total setiap Itemnya{total_Item}')

Total setiap ItemnyaItem
Juice       1171
Coffee      1165
Salad       1148
Cake        1139
Sandwich    1131
Smoothie    1096
Cookie      1092
Tea         1089
Unknown      969
Name: count, dtype: int64


## Quantity

hal pertama yang dilakukan pada feature Quantity yaitu mengecek data unique dan total missing values

In [9]:
unikQuantity = Data1['Quantity'].unique()
misvalQuantity = Data1['Quantity'].isna().sum()

print('Quantity Feauture')
print(f'Data Unik: \n{unikQuantity}\n')
print(f'Total Missing Value: \n{misvalQuantity}')

Quantity Feauture
Data Unik: 
['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN' nan]

Total Missing Value: 
138


dari data missing values diatas berjumlah 138 diisi dengan 'UNKNOWN', kemudian mengganti data 'UNKNOWN' dan 'ERROR' menjadi data null kembali

In [10]:
misvalQuantity1 = Data1['Quantity'] = Data1['Quantity'].fillna('UNKNOWN')
Data1['Quantity'] = Data1['Quantity'].apply(
    lambda x: np.nan if str(x).strip().upper() in ['UNKNOWN','ERROR']
    else x
)

print('Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown\n')
print('Data setelah di diisi dan digabungkan')
Data1['Quantity'].unique()

Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown

Data setelah di diisi dan digabungkan


array(['2', '4', '5', '3', '1', nan], dtype=object)

cek data null menjadi bertambah yang sebelumnya 138 menjadi 479 data

In [11]:
Data1['Quantity'].isna().sum()

np.int64(479)

hitung total dari masing-masing Quantitynya

In [12]:
total_Quantity = Data1['Quantity'].value_counts(dropna=False)
print(f'Total setiap Quantity nya: {total_Quantity}')

Total setiap Quantity nya: Quantity
5      2013
2      1974
4      1863
3      1849
1      1822
NaN     479
Name: count, dtype: int64


ubah type data pada kolom Quantity menjadi numerik

In [13]:
print('ubah type data menjadi numerik')
Data1['Quantity'] = pd.to_numeric(Data1['Quantity'])

ubah type data menjadi numerik


data null, diisi menggunakan median()

In [14]:
median_qty = Data1['Quantity'].median()
Data1['Quantity'] = Data1['Quantity'].fillna(median_qty)

total_Quantity_after = Data1['Quantity'].value_counts(dropna=False)
print(f'Total setiap Quantity nya setelah data null diisi dengan median: \n{total_Quantity_after}')

Total setiap Quantity nya setelah data null diisi dengan median: 
Quantity
3.0    2328
5.0    2013
2.0    1974
4.0    1863
1.0    1822
Name: count, dtype: int64


ubah kembali menjadi type data int agar mudah di analisa

In [15]:
Data1['Quantity'] = Data1['Quantity'].astype(int)

feature Quantity setelah dilakukan cleaning & missing values 
1. tidak memiliki data null
2. type data diubah menjadi int

In [16]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              10000 non-null  object
 2   Quantity          10000 non-null  int64 
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: int64(1), object(7)
memory usage: 625.1+ KB


## Price Per Unit

cek data unique dan missing values

In [17]:
unikPricePerUnit = Data1['Price Per Unit'].unique()
misvalPricePerUnit = Data1['Price Per Unit'].isna().sum()

print('Price Per Unit Feauture')
print(f'Data Unik: \n{unikPricePerUnit}\n')
print(f'Total Missing Value: \n{misvalPricePerUnit}')

Price Per Unit Feauture
Data Unik: 
['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' nan 'ERROR' 'UNKNOWN']

Total Missing Value: 
179


cek setiap 'Item' apakah memiliki harga yang berbeda

In [18]:
Data1.groupby('Item')['Price Per Unit'].nunique()

Item
Cake        3
Coffee      3
Cookie      3
Juice       3
Salad       3
Sandwich    3
Smoothie    3
Tea         3
Unknown     8
Name: Price Per Unit, dtype: int64

ditemukan ternyata pada setiap 'Item' berisi harga, UNKNWON, null dan ERROR

In [19]:
Data1.loc[Data1['Item'] == 'Coffee', 'Price Per Unit'].unique()

array(['2.0', nan, 'ERROR', 'UNKNOWN'], dtype=object)

menggabungkan data 'UNKNOWN' dan 'ERROR' menjadi null pada feature 'Price Per Unit'

In [20]:
Data1['Price Per Unit'] = Data1['Price Per Unit'].apply(
    lambda x: np.nan if str(x).strip().upper() in ['UNKNOWN','ERROR']
    else x
)

print('Gabungkan data UNKNOWN dan ERROR menjadi null\n')
print('Data setelah digabungkan')
Data1['Price Per Unit'].unique()

Gabungkan data UNKNOWN dan ERROR menjadi null

Data setelah digabungkan


array(['2.0', '3.0', '1.0', '5.0', '4.0', '1.5', nan], dtype=object)

setelah digabungkan masing-masing Item memiliki masing-masing 1 harga dan nilai null, namun pada Item 'Unknown' berisi 6 harga termasuk data null

In [21]:
afterGroup = Data1.groupby('Item')['Price Per Unit'].nunique()
print(afterGroup)

Item
Cake        1
Coffee      1
Cookie      1
Juice       1
Salad       1
Sandwich    1
Smoothie    1
Tea         1
Unknown     6
Name: Price Per Unit, dtype: int64


In [22]:
Data1.loc[Data1['Item'] == 'Unknown', 'Price Per Unit'].unique()

array(['3.0', '1.5', '2.0', '1.0', '5.0', nan, '4.0'], dtype=object)

cek harga unik yang dimiliki masing-masing Item

In [23]:
harga_produk = Data1.groupby('Item')['Price Per Unit'].unique()
harga_produk

Item
Cake                                 [3.0, nan]
Coffee                               [2.0, nan]
Cookie                               [1.0, nan]
Juice                                [3.0, nan]
Salad                                [5.0, nan]
Sandwich                             [4.0, nan]
Smoothie                             [4.0, nan]
Tea                                  [1.5, nan]
Unknown     [3.0, 1.5, 2.0, 1.0, 5.0, nan, 4.0]
Name: Price Per Unit, dtype: object

total missing values

In [24]:
Data1['Price Per Unit'].isna().sum()

np.int64(533)

sebelum mengisi missing values nya, ubah type data nya menjadi numerik

In [25]:
Data1['Price Per Unit'] = pd.to_numeric(Data1['Price Per Unit'])

kemudian isi missing values pada setiap Itemnya dengan harga yang tertera

In [26]:
# mengisi data Price Per Unit berdasarkan Item nya
Data1['Price Per Unit'] = Data1.groupby('Item')['Price Per Unit'] \
                              .transform(lambda x: x.fillna(x.median()))

setelah mengisi missing values, tidak ada lagi data null

In [27]:
harga_produk1 = Data1.groupby('Item')['Price Per Unit'].unique()
harga_produk1

Item
Cake                                 [3.0]
Coffee                               [2.0]
Cookie                               [1.0]
Juice                                [3.0]
Salad                                [5.0]
Sandwich                             [4.0]
Smoothie                             [4.0]
Tea                                  [1.5]
Unknown     [3.0, 1.5, 2.0, 1.0, 5.0, 4.0]
Name: Price Per Unit, dtype: object

cek apakah masih ada data null

In [28]:
Data1['Price Per Unit'].isna().sum()

np.int64(0)

cek total data pada feature 'Price Per Unit'

In [29]:
total_PricePerUnit = Data1['Price Per Unit'].value_counts()
print(total_PricePerUnit)

Price Per Unit
3.0    2598
4.0    2440
2.0    1284
5.0    1270
1.0    1209
1.5    1199
Name: count, dtype: int64


ubah tipe data menjadi float

In [30]:
Data1['Price Per Unit'] = Data1['Price Per Unit'].astype(float)

kolom 'Price Per Unit' tidak memiliki missing values dan type data dirubah menjadi float

In [31]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              10000 non-null  object 
 2   Quantity          10000 non-null  int64  
 3   Price Per Unit    10000 non-null  float64
 4   Total Spent       9827 non-null   object 
 5   Payment Method    7421 non-null   object 
 6   Location          6735 non-null   object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(1), int64(1), object(6)
memory usage: 625.1+ KB


## Total Spent

cek data unik dan missing values

In [32]:
unikTotalSpent = Data1['Total Spent'].unique()
misvalTotalSpent = Data1['Total Spent'].isna().sum()

print('Item Feauture')
print(f'Data Unik: \n{unikTotalSpent}\n')
print(f'Total Missing Value: \n{misvalTotalSpent}')

Item Feauture
Data Unik: 
['4.0' '12.0' 'ERROR' '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' nan 'UNKNOWN' '2.0' '1.0' '7.5' '4.5' '1.5']

Total Missing Value: 
173


menggabungkan data 'UNKNOWN' dan 'ERROR' menjadi data null

In [33]:
Data1['Total Spent'] = Data1['Total Spent'].apply(
    lambda x: np.nan if str(x).strip().upper() in ['UNKNOWN','ERROR']
    else x
)

print('Gabungkan data UNKNOWN dan ERROR menjadi null\n')
print('Data setelah digabungkan')
Data1['Total Spent'].unique()

Gabungkan data UNKNOWN dan ERROR menjadi null

Data setelah digabungkan


array(['4.0', '12.0', nan, '10.0', '20.0', '9.0', '16.0', '15.0', '25.0',
       '8.0', '5.0', '3.0', '6.0', '2.0', '1.0', '7.5', '4.5', '1.5'],
      dtype=object)

isi data null/missing values dengan hasil perkalian pada feature 'Quantity' dan 'Price Per Unit'

In [34]:
Data1['Total Spent'] = Data1['Total Spent'].fillna(
    Data1['Quantity'] * Data1['Price Per Unit']
)

print('Data null setelah diisi dengan perkalian Quantity dan Price Per Unit')
Data1['Total Spent'].unique()

Data null setelah diisi dengan perkalian Quantity dan Price Per Unit


array(['4.0', '12.0', 4.0, '10.0', '20.0', '9.0', '16.0', '15.0', '25.0',
       '8.0', '5.0', '3.0', '6.0', 12.0, 2.0, 3.0, '2.0', '1.0', '7.5',
       '4.5', 9.0, '1.5', 6.0, 15.0, 20.0, 7.5, 10.0, 25.0, 8.0, 1.0,
       16.0, 1.5, 4.5, 5.0], dtype=object)

cek missing values

In [35]:
Data1['Total Spent'].isna().sum()

np.int64(0)

ubah tipe data menjadi float

In [36]:
Data1['Total Spent'] = Data1['Total Spent'].astype(float)

feature 'Total Spent' tidak lagi memiliki missing values dan type data berubah menjadi float

In [37]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              10000 non-null  object 
 2   Quantity          10000 non-null  int64  
 3   Price Per Unit    10000 non-null  float64
 4   Total Spent       10000 non-null  float64
 5   Payment Method    7421 non-null   object 
 6   Location          6735 non-null   object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 625.1+ KB


## Payment Method

cek data unik dan missing values

In [38]:
unikPaymentMethod = Data1['Payment Method'].unique()
misvalPaymentMethod = Data1['Payment Method'].isna().sum()

print('Payment Method Feauture')
print(f'Data Unik: \n{unikPaymentMethod}\n')
print(f'Total Missing Value: \n{misvalPaymentMethod}')

Payment Method Feauture
Data Unik: 
['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR' nan]

Total Missing Value: 
2579


isi data null dengan 'UNKNOWN' kemudian gabung data 'UNKNWON' dan 'ERROR' menjadi 'Unknown'

In [39]:
misvalPaymentMethod = Data1['Payment Method'] = Data1['Payment Method'].fillna('UNKNOWN')
Data1['Payment Method'] = Data1['Payment Method'].apply(
    lambda x:'Unknown' if str(x) in ['UNKNOWN','ERROR']
    else str(x)
)

print('Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown\n')
print('Data setelah di diisi dan digabungkan')
Data1['Payment Method'].unique()

Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown

Data setelah di diisi dan digabungkan


array(['Credit Card', 'Cash', 'Unknown', 'Digital Wallet'], dtype=object)

total data unik pada masing-masing feature 'Payment Method'

**Catatan**: data 'Unknwon' pada feature 'Payment Method' terlalu banyak sehingga lebih baik ditetapkan menjadi data baru

In [40]:
total_PaymentMethod = Data1['Payment Method'].value_counts(dropna=False)
print(total_PaymentMethod)

Payment Method
Unknown           3178
Digital Wallet    2291
Credit Card       2273
Cash              2258
Name: count, dtype: int64


feature 'Payment Method' tidak terdapat missing values dan type data tetap object

In [41]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              10000 non-null  object 
 2   Quantity          10000 non-null  int64  
 3   Price Per Unit    10000 non-null  float64
 4   Total Spent       10000 non-null  float64
 5   Payment Method    10000 non-null  object 
 6   Location          6735 non-null   object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 625.1+ KB


## Location

cek data unik dan missing values

In [42]:
unikLocation = Data1['Location'].unique()
misvalLocation = Data1['Location'].isna().sum()

print('Location Feauture')
print(f'Data Unik: \n{unikLocation}\n')
print(f'Total Missing Value: \n{misvalLocation}')

Location Feauture
Data Unik: 
['Takeaway' 'In-store' 'UNKNOWN' nan 'ERROR']

Total Missing Value: 
3265


isi data null menjadi 'UNKNWON' kemudian gabung data 'UNKNOWN' dan 'ERROR' menjadi 'Unknown'

In [43]:
misvalLocation = Data1['Location'] = Data1['Location'].fillna('UNKNOWN')
Data1['Location'] = Data1['Location'].apply(
    lambda x:'Unknown' if str(x) in ['UNKNOWN','ERROR']
    else str(x)
)

print('Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown\n')
print('Data setelah di diisi dan digabungkan')
Data1['Location'].unique()

Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown

Data setelah di diisi dan digabungkan


array(['Takeaway', 'In-store', 'Unknown'], dtype=object)

total masing-masing data unik pada feature 'Location'

**Catatan**: data 'Unknwon' pada feature 'Location' terlalu banyak sehingga lebih baik ditetapkan menjadi data baru seperti pada feature sebelumnya

In [44]:
total_Location1 = Data1['Location'].value_counts(dropna=False)
print(total_Location1)

Location
Unknown     3961
Takeaway    3022
In-store    3017
Name: count, dtype: int64


tidak terdapat missing values pada feature 'Location' dan type data tetap object

In [45]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    10000 non-null  object 
 1   Item              10000 non-null  object 
 2   Quantity          10000 non-null  int64  
 3   Price Per Unit    10000 non-null  float64
 4   Total Spent       10000 non-null  float64
 5   Payment Method    10000 non-null  object 
 6   Location          10000 non-null  object 
 7   Transaction Date  9841 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 625.1+ KB


## Transaction Date

cek data unik dan missing values

In [46]:
unikTransactionDate = Data1['Transaction Date'].unique()
misvalTransactionDate = Data1['Transaction Date'].isna().sum()

print('Transaction Date Feauture')
print(f'Data Unik: \n{unikTransactionDate}\n')
print(f'Total Missing Value: \n{misvalTransactionDate}')

Transaction Date Feauture
Data Unik: 
['2023-09-08' '2023-05-16' '2023-07-19' '2023-04-27' '2023-06-11'
 '2023-03-31' '2023-10-06' '2023-10-28' '2023-07-28' '2023-12-31'
 '2023-11-07' 'ERROR' '2023-05-03' '2023-06-01' '2023-03-21' '2023-11-15'
 '2023-06-10' '2023-02-24' '2023-03-25' '2023-01-15' '2023-04-04'
 '2023-03-30' '2023-12-01' '2023-09-18' '2023-06-03' '2023-12-13'
 '2023-04-20' '2023-04-10' '2023-03-11' '2023-06-02' '2023-11-06'
 '2023-08-15' '2023-10-09' '2023-05-28' '2023-07-17' '2023-04-29'
 '2023-06-08' '2023-06-29' '2023-04-17' '2023-12-22' '2023-01-10'
 '2023-10-02' '2023-02-23' '2023-03-22' '2023-11-03' '2023-03-02'
 '2023-06-26' '2023-05-02' '2023-09-05' '2023-01-08' '2023-03-15'
 '2023-11-25' '2023-12-05' '2023-03-19' '2023-06-27' '2023-04-19'
 '2023-10-07' '2023-09-30' '2023-05-27' '2023-11-18' '2023-10-20'
 '2023-10-03' '2023-10-27' '2023-04-06' '2023-01-31' '2023-12-08'
 '2023-06-19' '2023-12-14' '2023-07-16' '2023-02-22' nan '2023-06-15'
 '2023-12-09' '2023-04-18'

data 'UNKNOWN' dan 'ERROR' diubah menjadi null

In [47]:
Data1['Transaction Date'] = Data1['Transaction Date'].apply(
    lambda x: np.nan if str(x).strip().upper() in ['UNKNOWN','ERROR']
    else x
)

print('Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown\n')
print('Data setelah di diisi dan digabungkan')
Data1['Transaction Date'].unique()

Isi data null dengan UNKNOWN dan gabungkan data UNKNOWN dan ERROR menjadi Unknown

Data setelah di diisi dan digabungkan


array(['2023-09-08', '2023-05-16', '2023-07-19', '2023-04-27',
       '2023-06-11', '2023-03-31', '2023-10-06', '2023-10-28',
       '2023-07-28', '2023-12-31', '2023-11-07', nan, '2023-05-03',
       '2023-06-01', '2023-03-21', '2023-11-15', '2023-06-10',
       '2023-02-24', '2023-03-25', '2023-01-15', '2023-04-04',
       '2023-03-30', '2023-12-01', '2023-09-18', '2023-06-03',
       '2023-12-13', '2023-04-20', '2023-04-10', '2023-03-11',
       '2023-06-02', '2023-11-06', '2023-08-15', '2023-10-09',
       '2023-05-28', '2023-07-17', '2023-04-29', '2023-06-08',
       '2023-06-29', '2023-04-17', '2023-12-22', '2023-01-10',
       '2023-10-02', '2023-02-23', '2023-03-22', '2023-11-03',
       '2023-03-02', '2023-06-26', '2023-05-02', '2023-09-05',
       '2023-01-08', '2023-03-15', '2023-11-25', '2023-12-05',
       '2023-03-19', '2023-06-27', '2023-04-19', '2023-10-07',
       '2023-09-30', '2023-05-27', '2023-11-18', '2023-10-20',
       '2023-10-03', '2023-10-27', '2023-04-06', '

ubah type data yang sebelumnya objec menjadi datetime(tanggal)

In [48]:
Data1['Transaction Date'] = pd.to_datetime(Data1['Transaction Date'])
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              10000 non-null  object        
 2   Quantity          10000 non-null  int64         
 3   Price Per Unit    10000 non-null  float64       
 4   Total Spent       10000 non-null  float64       
 5   Payment Method    10000 non-null  object        
 6   Location          10000 non-null  object        
 7   Transaction Date  9540 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 625.1+ KB


cek missing values

isi missing values dengan median

In [49]:
median_TransactionDate = Data1['Transaction Date'].median()
Data1['Transaction Date'] = Data1['Transaction Date'].fillna(median_qty)

cek missing values

In [50]:
Data1['Transaction Date'].isna().sum()

np.int64(0)

ubah lagi ke dalam datetime karena sebelumnya berubah menjadi numerik karena mengisi missing values

In [51]:
Data1['Transaction Date'] = pd.to_datetime(Data1['Transaction Date'])

tidak ada missing values pada feature 'Transaction Date' dan type data menjadi datetime

In [52]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              10000 non-null  object        
 2   Quantity          10000 non-null  int64         
 3   Price Per Unit    10000 non-null  float64       
 4   Total Spent       10000 non-null  float64       
 5   Payment Method    10000 non-null  object        
 6   Location          10000 non-null  object        
 7   Transaction Date  10000 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 625.1+ KB


**Menambahkan feature baru(Type, Category, Cost Per Item, Total Cost, Profit) untuk menambah insight**

## Type

kelompokkan setiap Item menjadi beberapa Type seperti Main Dish, Dessert, Cold Baverage, dan Hot Baverage

In [53]:
Jenis_Hidangan = {
    'Salad': 'Main Dish',
    'Sandwich': 'Main Dish',
    'Cake': 'Dessert',
    'Cookie': 'Dessert',
    'Smoothie': 'Cold Beverage',
    'Juice': 'Cold Beverage',
    'Coffee': 'Hot Beverage',
    'Tea': 'Hot Beverage',
    'Unknown': 'Unknown'
}

Data1['Type'] = Data1['Item'].map(Jenis_Hidangan) 

cek total data unik pada feature 'Type'

In [54]:
total_Jenis = Data1['Type'].value_counts(dropna=False)
print(total_Jenis)

Type
Main Dish        2279
Cold Beverage    2267
Hot Beverage     2254
Dessert          2231
Unknown           969
Name: count, dtype: int64


## Category

bagi setiap Item menjadi 2 kategori yaitu Food dan Drink

In [55]:
Kategori_Hidangan = {
    'Salad': 'Food',
    'Sandwich': 'Food',
    'Cake': 'Food',
    'Cookie': 'Food',
    'Smoothie': 'Drink',
    'Juice': 'Drink',
    'Coffee': 'Drink',
    'Tea': 'Drink',
    'Unknown': 'Unknown'
}

Data1['Category'] = Data1['Item'].map(Kategori_Hidangan) 

cek total data unik pada feature 'Category'

In [56]:
total_Kategori = Data1['Category'].value_counts(dropna=False)
print(total_Kategori)

Category
Drink      4521
Food       4510
Unknown     969
Name: count, dtype: int64


## Cost Per Item

tambahkan harga bahan baku pembuatan pada setiap Itemnya untuk menambahkan insight baru

In [57]:
cost_map = {
    'Salad': 2.0,
    'Sandwich': 1.6,
    'Cake': 1.05,
    'Cookie': 0.3,
    'Juice': 1.05,
    'Coffee': 0.5,
    'Smoothie': 1.6,
    'Tea': 0.3,
    'Unknown': 0
}

Data1['Cost Per Item'] = Data1['Item'].map(cost_map)
Data1['Cost Per Item'] = Data1['Cost Per Item'].astype(float)
Data1['Cost Per Item'].unique()

array([0.5 , 1.05, 0.3 , 2.  , 1.6 , 0.  ])

## Total Cost

buat perhitungan harga total bahan baku dari Item yang dibeli pada setiap transaksinya, dengan mengkalikan 'Quantity' dan 'Cost Per Item'

In [58]:
Data1['Total Cost'] = Data1['Cost Per Item'] * Data1['Quantity']
Data1['Total Cost'] = Data1['Total Cost'].round(2)
Data1['Total Cost'].unique()

array([ 1.  ,  4.2 ,  1.2 ,  4.  ,  8.  ,  0.  ,  6.4 , 10.  ,  3.2 ,
        1.5 ,  6.  ,  1.6 ,  2.1 ,  5.25,  4.8 ,  2.  ,  0.6 ,  3.15,
        0.9 ,  0.3 ,  0.5 ,  2.5 ,  1.05])

## Profit

feature baru 'Profit' digunakan untuk mengetahui keuntungan bersih dari masing-masing Item yang dijual

In [59]:
Data1['Profit'] = Data1['Total Spent'] - Data1['Total Cost']
Data1['Profit'].unique()

array([ 3.  ,  7.8 ,  2.8 ,  6.  , 12.  ,  9.  ,  9.6 , 15.  ,  4.8 ,
        3.5 ,  2.4 ,  3.9 ,  9.75, 15.2 ,  7.2 , 10.  ,  2.  ,  5.  ,
        1.4 ,  5.85, 25.  ,  1.1 , -0.15,  2.85,  0.7 ,  1.  ,  1.5 ,
        3.6 ,  4.5 ,  2.1 ,  7.5 ,  1.2 ,  1.95, 16.  , 20.  ,  8.  ,
       19.  , 11.2 ,  4.  ,  8.85,  6.5 ,  0.1 ,  2.5 , -0.8 ,  3.2 ,
        4.1 , 14.  ,  6.6 ,  0.5 , -1.  ,  0.6 ,  8.5 ,  5.1 , 11.85,
        3.1 ])

In [60]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    10000 non-null  object        
 1   Item              10000 non-null  object        
 2   Quantity          10000 non-null  int64         
 3   Price Per Unit    10000 non-null  float64       
 4   Total Spent       10000 non-null  float64       
 5   Payment Method    10000 non-null  object        
 6   Location          10000 non-null  object        
 7   Transaction Date  10000 non-null  datetime64[ns]
 8   Type              10000 non-null  object        
 9   Category          10000 non-null  object        
 10  Cost Per Item     10000 non-null  float64       
 11  Total Cost        10000 non-null  float64       
 12  Profit            10000 non-null  float64       
dtypes: datetime64[ns](1), float64(5), int64(1), object(6)
memory usage: 1015.8+ K

dari hasil data unik di atas terdapat bebarapa data yang salah perhitungan, maka dilakukan penghapusan data yang tidak sesusai

In [61]:
Data1 = Data1[Data1['Profit'] >= 0]
Data1['Profit'].unique()

array([ 3.  ,  7.8 ,  2.8 ,  6.  , 12.  ,  9.  ,  9.6 , 15.  ,  4.8 ,
        3.5 ,  2.4 ,  3.9 ,  9.75, 15.2 ,  7.2 , 10.  ,  2.  ,  5.  ,
        1.4 ,  5.85, 25.  ,  1.1 ,  2.85,  0.7 ,  1.  ,  1.5 ,  3.6 ,
        4.5 ,  2.1 ,  7.5 ,  1.2 ,  1.95, 16.  , 20.  ,  8.  , 19.  ,
       11.2 ,  4.  ,  8.85,  6.5 ,  0.1 ,  2.5 ,  3.2 ,  4.1 , 14.  ,
        6.6 ,  0.5 ,  0.6 ,  8.5 ,  5.1 , 11.85,  3.1 ])

# Analisis

cek data unik pada setiap feature

In [62]:
for dtype, col in (list(zip(Data1.dtypes, Data1.columns))):
    print(col, dtype)
    print(Data1[col].unique())
    print()

Transaction ID object
['TXN_1961373' 'TXN_4977031' 'TXN_4271903' ... 'TXN_5255387' 'TXN_7695629'
 'TXN_6170729']

Item object
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'Unknown' 'Sandwich' 'Juice'
 'Tea']

Quantity int64
[2 4 5 3 1]

Price Per Unit float64
[2.  3.  1.  5.  4.  1.5]

Total Spent float64
[ 4.  12.  10.  20.   9.  16.  15.  25.   8.   5.   3.   6.   2.   1.
  7.5  4.5  1.5]

Payment Method object
['Credit Card' 'Cash' 'Unknown' 'Digital Wallet']

Location object
['Takeaway' 'In-store' 'Unknown']

Transaction Date datetime64[ns]
<DatetimeArray>
['2023-09-08 00:00:00', '2023-05-16 00:00:00', '2023-07-19 00:00:00',
 '2023-04-27 00:00:00', '2023-06-11 00:00:00', '2023-03-31 00:00:00',
 '2023-10-06 00:00:00', '2023-10-28 00:00:00', '2023-07-28 00:00:00',
 '2023-12-31 00:00:00',
 ...
 '2023-08-01 00:00:00', '2023-01-20 00:00:00', '2023-11-11 00:00:00',
 '2023-02-13 00:00:00', '2023-07-30 00:00:00', '2023-02-17 00:00:00',
 '2023-05-20 00:00:00', '2023-11-05 00:00:00', '2023-0

hapus data 'Unknown' pada feature 'Item' karena tidak digunakan untuk analisa

In [63]:
Data1 = Data1[Data1['Item'] != 'Unknown']
Data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8984 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    8984 non-null   object        
 1   Item              8984 non-null   object        
 2   Quantity          8984 non-null   int64         
 3   Price Per Unit    8984 non-null   float64       
 4   Total Spent       8984 non-null   float64       
 5   Payment Method    8984 non-null   object        
 6   Location          8984 non-null   object        
 7   Transaction Date  8984 non-null   datetime64[ns]
 8   Type              8984 non-null   object        
 9   Category          8984 non-null   object        
 10  Cost Per Item     8984 non-null   float64       
 11  Total Cost        8984 non-null   float64       
 12  Profit            8984 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(1), object(6)
memory usage: 982.6+ KB


cek tahun dari Transaction Date

In [64]:
Data1['Transaction Date'].dt.year.unique()

array([2023, 1970], dtype=int32)

ditemukan ternyata terdapat data yang outlier yaitu 1970 maka dari itu dilakukan penghapusan data yang memiliki tahun 1970

In [65]:
Data1['Transaction Date'] = pd.to_datetime(Data1['Transaction Date'], errors='coerce')
Data1 = Data1[Data1['Transaction Date'].dt.year != 1970]
Data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8568 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    8568 non-null   object        
 1   Item              8568 non-null   object        
 2   Quantity          8568 non-null   int64         
 3   Price Per Unit    8568 non-null   float64       
 4   Total Spent       8568 non-null   float64       
 5   Payment Method    8568 non-null   object        
 6   Location          8568 non-null   object        
 7   Transaction Date  8568 non-null   datetime64[ns]
 8   Type              8568 non-null   object        
 9   Category          8568 non-null   object        
 10  Cost Per Item     8568 non-null   float64       
 11  Total Cost        8568 non-null   float64       
 12  Profit            8568 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(1), object(6)
memory usage: 937.1+ KB


ubah nama pada kolom 'Price Per Unit' dan 'Total Spent'

In [66]:
Data1.rename(columns={
    'Price Per Unit':'Price Per Item',
    'Total Spent':'Total Price'
}, inplace=True)

# Verify column names after renaming
print(Data1.columns)

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Item', 'Total Price',
       'Payment Method', 'Location', 'Transaction Date', 'Type', 'Category',
       'Cost Per Item', 'Total Cost', 'Profit'],
      dtype='object')


In [67]:
Data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8568 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    8568 non-null   object        
 1   Item              8568 non-null   object        
 2   Quantity          8568 non-null   int64         
 3   Price Per Item    8568 non-null   float64       
 4   Total Price       8568 non-null   float64       
 5   Payment Method    8568 non-null   object        
 6   Location          8568 non-null   object        
 7   Transaction Date  8568 non-null   datetime64[ns]
 8   Type              8568 non-null   object        
 9   Category          8568 non-null   object        
 10  Cost Per Item     8568 non-null   float64       
 11  Total Cost        8568 non-null   float64       
 12  Profit            8568 non-null   float64       
dtypes: datetime64[ns](1), float64(5), int64(1), object(6)
memory usage: 937.1+ KB


simpan data yang telah dibersihkan ke bentuk csv yang nantinya akan divisualisasikan menggunakan Power BI

In [68]:
Data1 = Data1.to_csv('Clean_Cafe_Sales.csv')